scipy/cluster/hierarchy.py

```diff
@@ -163,7 +163,7 @@ def _warning(s):
 def int_floor(arr, xp):
     # numpy.array_api is strict about not allowing `int()` on a float array.
     # That's typically not needed, here it is - so explicitly convert
-    return int(xp.astype(arr, xp.int64))
+    return int(xp.astype(xp.asarray(arr), xp.int64))
```

Comment:
```suggestion
    return int(xp.asarray(arr, dtype=xp.int64))
```

Reply:
Also, why is this change needed? Does `xp` not come from `array_namespace(arr)`?

Reply:
I'll verify, but IIRC `arr` can be either a 1-D array or a scalar, and the latter didn't work.

Reply:
Actually, it was a scalar only because we were bypassing `array_api_compat`; now that that's fixed this change is no longer needed it looks like.

Reply:
The problem stems from indexing; arrays are converted correctly but then `arr = row[0]` may produce a scalar again - and `int_floor` is called from multiple places.

Reply:
No, it doesn't quite, because (from the [`asarray` docs](https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html)):

<img width="502" alt="image" src="https://github.com/scipy/scipy/assets/98330/d60c345a-f3b3-4bd6-ab48-44af34a0eddd">

For unsafe casting from float to int, `astype` should be correct here.

Reply:
Ah yes, I remember that pattern now that you mention it!